## JUMIA-PRODUCTS-RECOMMENDATION-AND-PRODUCT-REVIEW-SYSTEM

### Overview

The growth of e-commerce in Kenya has been exponential, with platforms like Jumia attracting large numbers of users. However, users frequently encounter challenges in locating relevant products due to expansive and sometimes disorganized product catalogs. Moreover, the limited use of recommendations and unreliable review systems further complicate the shopping experience, leading to reduced user satisfaction and lower conversion rates. To address these issues, we propose a recommendation and sentiment analysis system that leverages interaction history and user feedback. This system aims to provide more relevant recommendations and insights from customer sentiments, enabling Jumia to better serve its users and enhance its retail offerings.

### Problem Statement

The problem of finding relevant products on Jumia’s extensive catalog significantly affects customer satisfaction and conversion rates. Current limitations include:  

* Inadequate product recommendations based on user behavior, which leads to a lack of personalization.
* An unreliable rating and review system, leaving customers with limited insights into product quality.
* Limited retailer insights into customer sentiment, hindering their ability to refine offerings and optimize marketing strategies.
  
These issues reduce user satisfaction, decrease sales, and limit the potential for enhanced customer-retailer relationships on Jumia.

### Stakeholders

**Primary Users (Consumers)**: Customers who frequently shop on Jumia and need relevant recommendations to streamline their product search and improve purchase decisions.

**Retailers/Sellers**: Businesses and individual sellers on Jumia who seek insights into customer preferences and sentiment to better tailor their product offerings and marketing strategies.

**Jumia Management**: The platform administrators who aim to improve customer satisfaction, engagement, and conversion rates through improved site functionality.

### Objectives

* Develop a Recommendation System:

Create an intelligent recommendation engine that uses customer interaction history, preferences, and similar behavior patterns to provide relevant product recommendations.

* Implement Sentiment Analysis:

Extract sentiments from customer reviews to offer meaningful feedback, categorizing it as positive, negative, or neutral. This feedback will help refine recommendations and improve customer trust in the platform.

* Enhance User Experience:

Improve customer satisfaction by reducing the time and effort needed to find relevant products, potentially increasing conversion rates.

* Provide Insightful Retailer Analytics:

Equip sellers with sentiment analysis reports and recommendation patterns to refine product and marketing offerings based on customer preferences.

* Support Kenyan E-commerce Growth:

Contribute to the development of the e-commerce sector in Kenya by addressing key pain points in product discoverability and review reliability on Jumia.

### Expected Benefits

* **For Customers**: More personalized recommendations, improved trust through reliable sentiment analysis, and an overall enhanced shopping experience.

* **For Retailers**: Access to better customer insights and sentiment data, allowing them to optimize product listings, improve customer engagement, and refine marketing strategies.

* **For Jumia**: Increased user satisfaction and conversion rates, resulting in stronger brand loyalty and a more competitive edge in the Kenyan e-commerce market.

### Data Understanding

We scraped data from [jumia](https://www.jumia.co.ke/) which is an online shop with more than 15 categories of products but decided to only have 15 categories

### Exploratory Data Ananlysis

Importing relevant libraries

In [3]:
#import files
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split,cross_val_score, GridSearchCV
from imblearn.over_sampling import SMOTE 
from scipy import stats as stats
from imblearn.pipeline import Pipeline as imbpipe
from sklearn.metrics import accuracy_score,f1_score,recall_score,precision_score,\
roc_curve,roc_auc_score,classification_report, auc, ConfusionMatrixDisplay
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from scipy.stats import randint
from xgboost import XGBClassifier
from sklearn.pipeline import Pipeline
from scipy import stats
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

import warnings
warnings.filterwarnings('ignore')

### Data Understanding

Loading the scraped dataset

In [4]:
# Load the data into a dataframe and read the first five rows
df = pd.read_csv('./Data/combined_dta.csv', encoding='ISO-8859-1')

df.head()

web-scraper-order                          web-scraper-start-url  \
0      1730407108-1  https://www.jumia.co.ke/computer-accessories/   
1      1730407112-2  https://www.jumia.co.ke/computer-accessories/   
2      1730407112-3  https://www.jumia.co.ke/computer-accessories/   
3      1730407114-4  https://www.jumia.co.ke/computer-accessories/   
4      1730407116-5  https://www.jumia.co.ke/computer-accessories/   

                                             product  \
0  Black Friday dealLaptop Skin Wrap Vinyl from 1...   
1  Black Friday deal9"-17" Laptop Cooling Stand P...   
2  Black Friday deal9"-17" Laptop Cooling Stand P...   
3  Black Friday dealLexar 2.5" SATA 256GB Laptop ...   
4      WD External Portable Hard Drive-2TBKSh 15,000   

                                        product-href  \
0  https://www.jumia.co.ke/generic-laptop-skin-wr...   
1  https://www.jumia.co.ke/9-17-laptop-cooling-st...   
2  https://www.jumia.co.ke/9-17-laptop-cooling-st...   
3  https://www.jumia.co.ke/lexar-2.5-sata-256gb-l...   
4  https://www.jumia.co.ke/wd-external-portable-h...   

                                        product_name  \
0  Laptop Skin Wrap Vinyl from 12-17.0 inches Lap...   
1                    9"-17" Laptop Cooling Stand Pad   
2                    9"-17" Laptop Cooling Stand Pad   
3                   Lexar 2.5" SATA 256GB Laptop SSD   
4                WD External Portable Hard Drive-2TB   

                                               image              price  \
0  https://ke.jumia.is/unsafe/fit-in/680x680/filt...  KSh 699 - KSh 797   
1  https://ke.jumia.is/unsafe/fit-in/680x680/filt...          KSh 3,325   
2  https://ke.jumia.is/unsafe/fit-in/680x680/filt...          KSh 3,325   
3  https://ke.jumia.is/unsafe/fit-in/680x680/filt...          KSh 3,799   
4  https://ke.jumia.is/unsafe/fit-in/680x680/filt...         KSh 15,000   

   brand link                                          link-href  see_all  \
0    NaN  NaN  https://www.jumia.co.ke/computer-accessories/?...      NaN   
1    NaN  NaN  https://www.jumia.co.ke/computer-accessories/?...  See All   
2    NaN  NaN  https://www.jumia.co.ke/computer-accessories/?...  See All   
3  Lexar  NaN  https://www.jumia.co.ke/computer-accessories/?...      NaN   
4     WD  NaN  https://www.jumia.co.ke/computer-accessories/?...      NaN   

                                        see_all-href review_title  \
0                                                NaN          NaN   
1  https://www.jumia.co.ke/catalog/productratings...         OKAY   
2  https://www.jumia.co.ke/catalog/productratings...         okay   
3                                                NaN          NaN   
4                                                NaN          NaN   

                                              review     ratings   user_name  \
0                                                NaN         NaN         NaN   
1                               Working, feels cheap  3 out of 5  by WILFRED   
2  A bit big for a regular laptop bag... The fanc...  4 out of 5  by Michael   
3                                                NaN         NaN         NaN   
4                                                NaN         NaN         NaN   

         date see_all_link see_all_link-href   Category  
0         NaN          NaN               NaN  Computing  
1  28/04/2022          NaN               NaN  Computing  
2    8/7/2021          NaN               NaN  Computing  
3         NaN          NaN               NaN  Computing  
4         NaN          NaN               NaN  Computing

In [5]:
# Get the number of rows and columns
print(df.shape, '\n')

print(f'The scraped data have {df.shape[0]} rows and {df.shape[1]} columns')

(60464, 20) 

The scraped data have 60464 rows and 20 columns


To arrive at this number of rows, we ensured that we scraped as much data as possible both from the product categories and sub-categories despite being blocked from sourcing most of it but we believe that this dataset will be helpful in achieving our objectives

In [6]:
# Let's look at the data information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60464 entries, 0 to 60463
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   web-scraper-order      60464 non-null  object
 1   web-scraper-start-url  60464 non-null  object
 2   product                60464 non-null  object
 3   product-href           60464 non-null  object
 4   product_name           60459 non-null  object
 5   image                  60330 non-null  object
 6   price                  60330 non-null  object
 7   brand                  39584 non-null  object
 8   link                   28139 non-null  object
 9   link-href              36020 non-null  object
 10  see_all                40431 non-null  object
 11  see_all-href           40425 non-null  object
 12  review_title           37705 non-null  object
 13  review                 40201 non-null  object
 14  ratings                37618 non-null  object
 15  user_name          

In [7]:
print(f'We have {df.shape[1]} columns in our dataset', '\n')

print('The columns are:')

df.columns

We have 20 columns in our dataset 

The columns are:


Index(['web-scraper-order', 'web-scraper-start-url', 'product', 'product-href',
       'product_name', 'image', 'price', 'brand', 'link', 'link-href',
       'see_all', 'see_all-href', 'review_title', 'review', 'ratings',
       'user_name', 'date', 'see_all_link', 'see_all_link-href', 'Category'],
      dtype='object')

To have better understanding of what each column entails and what they mean, have a look at the `column_description.txt` file in the `Data` folder

### Data Cleaning

In [8]:
# Function to calculate the percentage of null values in every column

def null_percentages(df):
    
    # Calculate the number of missing values and their percentage
    null_values = df.isnull().sum()
    null_percentage = (null_values / df.shape[0]) * 100
    
    # Create a DataFrame with the results
    missing_data = pd.DataFrame({'Null Values': null_values, 'Percentage': null_percentage})

    # Filter out columns with no missing values and sort by 'Missing Values'
    missing_data = missing_data[missing_data['Null Values'] > 0].sort_values(by='Null Values', ascending=False)

    return missing_data

In [9]:
null_percentages(df)

Null Values  Percentage
see_all_link             60413   99.915652
see_all_link-href        60413   99.915652
link                     32325   53.461564
link-href                24444   40.427362
ratings                  22846   37.784467
user_name                22846   37.784467
date                     22846   37.784467
review_title             22759   37.640580
brand                    20880   34.532945
review                   20263   33.512503
see_all-href             20039   33.142035
see_all                  20033   33.132112
image                      134    0.221619
price                      134    0.221619
product_name                 5    0.008269

We have lots of columns with over **35%** null values and so, we are going to of how to deal with the null values and at the same time maintain as much data as possible